[[ Bug 22121 ]] Implement delete undo for widgets

montegoulding opened the pull request:
This patch ensures that when widgets are added to the undo queue when the selected
objects are deleted that the undo will rebind the widget correctly.

runrevmark commented on engine/src/widget.cpp:
```diff
@@ -776,6 +777,41 @@ Boolean MCWidget::del(bool p_check_flag)
+    
+    if (m_rep == nullptr)
+    {
+        m_rep = kMCNull;
```
m_rep isn't an auto valueref so you need to MCValueRetain(kMCNull) here

runrevmark commented:
Looks good so far @montegoulding let me know when you've added tests and I'll run it through some memory leak checks (just to make sure there isn't something we've missed in the interplay between undo and redo).

runrevmark commented on engine/src/widget.cpp:
```diff
@@ -776,6 +777,41 @@ Boolean MCWidget::del(bool p_check_flag)
+Boolean MCWidget::delforundo(bool p_check_flag)
+{
+    // Make the widget generate a rep.
+    if (m_widget != nil && m_rep == nullptr)
```
I suggest saving the output of OnSave into a t_rep autocontainer - then only changing m_rep if `del` returns true - otherwise m_rep will be set whilst the control is bound.

runrevmark replied:
(Basically there should be no changed to the MCWidget's instance vars if del() returns false)

runrevmark commented on engine/src/widget.cpp:
```diff
@@ -776,6 +777,50 @@ Boolean MCWidget::del(bool p_check_flag)
+        }
+    }
+    
+    if (del(p_check_flag))
```
A better structure here would be:
```
if (!del(p_check_flags))
{
  return false;
}

m_rep = t_rep.Take();
return true;
```

montegoulding committed:
[[ Bug 22121 ]] Implement delete undo for widgets

This patch ensures that when widgets are added to the undo queue when the selected
objects are deleted that the undo will rebind the widget correctly. This special
case of deleting a widget but leaving it in a state where no data loss will occur if
it is replaced with undo is handled by a new method `delforundo`.

This method ensures that the widget on the undo queue is in its correct unbound state
with an updated rep if the widget was bound. When the undo operation is a replace of
a deleted widget then the widget is re-bound. When the undo operation is a delete of
a replaced widget then the `delforundo` method is again called.

runrevmark commented:
I think this perhaps deserves a slightly more detailed commit message, don't you @montegoulding ? 😃

montegoulding commented:
@runrevmark I have fleshed it out a bit describing the change. Is there anything more you wanted covered in the msg?

runrevmark commented:
@livecode-vulcan review ok 2208e44

livecode-vulcan commented:
💙 review by @runrevmark ok 2208e44cccd2f86b096dd05d8534700704edd054

livecode-vulcan commented:
😎 test success 2208e44cccd2f86b096dd05d8534700704edd054

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/584)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/576)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/576)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/578)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/580)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2103)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/325)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/325)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2129)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2119)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2118)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2152)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2092)

runrevmark commented:
Just restarting the travis builder which failed - to double-check it was a transient failure.
